core/domain/stats_jobs_one_off.py

```diff
@@ -18,14 +18,466 @@
+                for state_hit_event in state_hit_event_dicts_sessioned[:-1]:
+                    state_stats_mapping[state_hit_event[
+                        'state_name']].num_completions_v1 += 1
+                if session_id in completed_session_ids:
+                    state_stats_mapping[state_hit_event_dicts_sessioned[-1][
+                        'state_name']].num_completions_v1 += 1
+
+            # Compute total_answers_count and useful_feedback_count.
+            for value in state_answer_events_for_this_version:
+                for answer in value['submitted_answer_list']:
+                    state_stats_mapping[value[
+                        'state_name']].total_answers_count_v1 += 1
+                    if answer['classification_categorization'] != (
+                            exp_domain.DEFAULT_OUTCOME_CLASSIFICATION):
+                        state_stats_mapping[value[
+                            'state_name']].useful_feedback_count_v1 += 1
+
+            # Now that the stats for the exploration's version are computed,
+            # calculate num_actual_starts. To compute this, we take the max of
+            # the first hit counts of all states that the initial state leads
+            # to. The max is taken instead of sum because the sum of
+            # first_hit_counts will sometimes end up counting a single learner
+            # twice. The max of both first hit counts ensures that we are
+            # counting only unique learners who traversed past the initial
+            # state.
+            max_first_hit_from_init_state = 0
+            for answer_group in versioned_exploration.states[versioned_exploration.init_state_name].interaction.answer_groups: # pylint: disable=line-too-long
+                if answer_group.outcome.dest != (
+                        versioned_exploration.init_state_name):
+                    if state_stats_mapping[
+                            answer_group.outcome.dest].first_hit_count_v1 > (
+                                max_first_hit_from_init_state):
+                        max_first_hit_from_init_state = state_stats_mapping[
+                            answer_group.outcome.dest].first_hit_count_v1
+            if max_first_hit_from_init_state == 0:
+                num_actual_starts = num_starts
+            else:
+                num_actual_starts = max_first_hit_from_init_state
+
+            # Check if model already exists. If it does, update it, otherwise
+            # create a fresh ExplorationStatsModel instance.
+            exploration_stats_model = (
+                stats_models.ExplorationStatsModel.get_model(exp_id, version))
+            # Handling exploration reverts.
+            if is_revert:
+                old_exploration_stats = (
+                    stats_services.get_exploration_stats_by_id(
+                        exp_id, revert_to_version))
+                num_starts = old_exploration_stats.num_starts_v1
```

Comment:
Er, wait a sec. If we're just going to throw away all the data above (in this case) then why compute it in the first place?

Consider extracting some of the above into subsidiary functions and calling them only if needed. Then you can do: `if is_revert: ... else: ...` more cleanly here, where the subsidiary functions are only called in the else branch.

Reply:
Yes, that sounds reasonable.